[MLIR][SparseTensor] Loop ordering strategy infrastructure (flag)

gmalasan opened the pull request:
As discussed before, this PR adds the basic infrastructure/boiler plate for loop ordering strategies to be implemented. 

If this looks ok, I wanted to also mention some of the heuristics that I would implement next, if they sound reasonable to you guys:
- Parallel first : prioritize parallel loops over reduction loops
- Dense outer : prioritize the most dense loops first
- Sparse outer : the opposite, potentially useful in some cases?

There is another that I am considering, stride/memory aware, which would prioritize loops with better stride patterns (like sequential or linear). Not sure how well this carries over to Sparse Tensor though. Are there any ideas/heuristics that I should definitely try to implement?

As we discussed, I will try to incrementally add heuristics. Sorry for the delay on my end, and thank you so much for the feedback!

gmalasan committed:
[MLIR][SparseTensor] Loop ordering strategy infrastructure (flag)

gmalasan committed:
[MLIR][SparseTensor] Fixed up the rest of the boilerplate code, strategy fully available in the topoSort method in IterationGraphSorter

aartbik commented on mlir/include/mlir/Dialect/SparseTensor/Transforms/Passes.h:
```diff
@@ -55,6 +55,15 @@ enum class SparseEmitStrategy {
 
+namespace sparse_tensor {
+
+/// Selects between different loop ordering strategies for sparse tensor 
```
Please pay attention to style. Period at end of comments (unless inline).
Also, in line with the other comments, I would say (since this is a struct not a method)

// Defines a strategy for loop ordering during sparse code generation.

aartbik commented on mlir/include/mlir/Dialect/SparseTensor/Transforms/Passes.td:
```diff
@@ -81,6 +81,11 @@ def SparseReinterpretMap : Pass<"sparse-reinterpret-map", "ModuleOp"> {
+       "mlir::sparse_tensor::LoopOrderingStrategy::kDefault",
+       "Set the loop ordering strategy for sparse tensor dialect", [{llvm::cl::values(
+         clEnumValN(mlir::sparse_tensor::LoopOrderingStrategy::kDefault, "default",
+                    "Default strategy (current behavior)"))}]>,
```
current behavior is very time dependent ;-)

aartbik commented on mlir/include/mlir/Dialect/SparseTensor/Transforms/Passes.td:
```diff
@@ -81,6 +81,11 @@ def SparseReinterpretMap : Pass<"sparse-reinterpret-map", "ModuleOp"> {
                     "Run on operations expect linalg.generic (e.g., foreach)"))}]>,
+    Option<"loopOrderingStrategy", "loop-ordering-strategy", "mlir::sparse_tensor::LoopOrderingStrategy",
+       "mlir::sparse_tensor::LoopOrderingStrategy::kDefault",
+       "Set the loop ordering strategy for sparse tensor dialect", [{llvm::cl::values(
```
it is not really for sparse tensor dialect, but for the sparse code generation

aartbik commented on mlir/lib/Dialect/SparseTensor/Transforms/SparseReinterpretMap.cpp:
```diff
@@ -421,7 +423,8 @@ struct GenericOpScheduler : public OpRewritePattern<linalg::GenericOp> {
       return failure();
 
-    auto scheduler = IterationGraphSorter::fromGenericOp(linalgOp);
+    // Pass strategy to IterationGraphSorter
```
period at end

aartbik commented on mlir/lib/Dialect/SparseTensor/Transforms/Utils/IterationGraphSorter.cpp:
```diff
@@ -100,7 +100,15 @@ AffineMap IterationGraphSorter::topoSort() {
     auto &it = !parIt.empty() ? parIt : redIt;
-    auto src = it.back();
+    
+    // Select loop based on strategy
```
period at end

aartbik commented on mlir/lib/Dialect/SparseTensor/Transforms/Utils/IterationGraphSorter.h:
```diff
@@ -42,8 +43,9 @@ enum class SortMask : unsigned {
   /// Factory method that construct an iteration graph sorter
-  /// for the given linalg.generic operation.
-  static IterationGraphSorter fromGenericOp(linalg::GenericOp genericOp);
+  /// for the given linalg.generic operation with a specific strategy.
```
with a specific loop ordering strategy

gmalasan committed:
[MLIR][SparseTensor] Fixed PR feedback about style

gmalasan committed:
Merge branch 'main' into sparse-tensor-loop-ordering-infrastructure

gmalasan committed:
Merge branch 'main' into sparse-tensor-loop-ordering-infrastructure

gmalasan commented:
Hey @aartbik, thank you so much for the feedback! I believe I've addressed the issues you pointed out, I'd love to know if there's anything else I should fix. 

Otherwise, I'd like to ask if what I mentioned earlier for my next steps make sense, implementing a parallel first, dense outer, and a sparse outer heuristic.

I would love to hear your thoughts about these strategies, or honestly if there are any heuristics you think are worth implementing. Thanks again!

gmalasan committed:
Merge branch 'main' into sparse-tensor-loop-ordering-infrastructure

gmalasan committed:
Merge branch 'main' into sparse-tensor-loop-ordering-infrastructure

aartbik reviewed (commented):
two last nits. 

as for next steps, once this PR goes in, please start incrementally adding new strategies, but make sure to provide evidence of usefulness of each, and avoid hard coding values that are very specific for the experiments you conducted (or at least document that properly)

gmalasan committed:
[MLIR][SparseTensor] Comment style fixes

gmalasan committed:
Merge branch 'sparse-tensor-loop-ordering-infrastructure' of https://github.com/gmalasan/llvm-project into sparse-tensor-loop-ordering-infrastructure
Merging other changes with comment style fixes

gmalasan committed:
[MLIR][SparseTensor] Missed comment style fix in Passes.h

gmalasan committed:
Merge branch 'main' into sparse-tensor-loop-ordering-infrastructure

gmalasan committed:
Merge branch 'main' into sparse-tensor-loop-ordering-infrastructure

aartbik commented:
Wait, are you continuing with new development in this PR? The idea was that we reviewed and approved the flag change, so that can go into main. After that, you send out new PRs (with much smaller deltas) with the new heuristics that we discuss in turn then.

If you just build on top of this PR we again get very huge differences and I have to constantly filter what is new and old?

gmalasan commented:
Sorry about that, I reset this branch to the last state that you approved and I'm opening a new PR for this heuristic. Thank you for letting me know the process, it definitely makes sense not to do so much in one PR.

gmalasan committed:
Merge branch 'main' into sparse-tensor-loop-ordering-infrastructure

aartbik reviewed (approved):
Great, yes, please submit this and then make a new PR for the first new heuristic, which we will discuss there!

gmalasan committed:
Merge branch 'main' into sparse-tensor-loop-ordering-infrastructure

gmalasan committed:
Merge branch 'main' into sparse-tensor-loop-ordering-infrastructure

gmalasan committed:
Apply clang-format to C++ files for loop ordering infrastructure

gmalasan committed:
Merge branch 'sparse-tensor-loop-ordering-infrastructure' of https://github.com/gmalasan/llvm-project into sparse-tensor-loop-ordering-infrastructure
Merging changes from main with local formatting changes

gmalasan committed:
Merge branch 'main' into sparse-tensor-loop-ordering-infrastructure

gmalasan committed:
Fix clang-format issues in SparseTensorPasses.cpp

gmalasan committed:
Merge branch 'sparse-tensor-loop-ordering-infrastructure' of https://github.com/gmalasan/llvm-project into sparse-tensor-loop-ordering-infrastructure
Merging changes in infrastructure commit formatting with upstream
changes

gmalasan committed:
Merge branch 'main' into sparse-tensor-loop-ordering-infrastructure

gmalasan commented:
I believe the formatting issues are now fixed.

gmalasan committed:
Merge branch 'main' into sparse-tensor-loop-ordering-infrastructure

aartbik committed:
Merge branch 'main' into sparse-tensor-loop-ordering-infrastructure

aartbik committed:
Merge branch 'main' into sparse-tensor-loop-ordering-infrastructure
